File: clients/flutter/lib/src/transport/transport.dart
import 'package:http/http.dart' as http;

import '../input.dart';
import '../response.dart';

/// The canonical, protocol-independent input handed to a [Transport].
class TransformRequest {
  /// Creates a transform request.
  const TransformRequest({
    required this.baseUrl,
    required this.headers,
    required this.input,
    required this.operations,
    required this.timeout,
    required this.httpClient,
  });

  /// The server base URL, matching the chosen transport.
  final String baseUrl;

  /// Default headers sent with every request.
  final Map<String, String> headers;

  /// The image source.
  final ImageInput input;

  /// The canonical operation set.
  final Map<String, Object?> operations;

  /// The per-request timeout.
  final Duration timeout;

  /// The HTTP client used by the REST transport and for [ImageInput.url] sources.
  final http.Client httpClient;
}

/// Speaks one of the Sharptown protocols. Each implementation accepts the same
/// [TransformRequest] and returns the same [TransformResponse], so swapping one transport
/// for another never changes calling code.
abstract interface class Transport {
  /// Runs a single transform over this transport.
  Future<TransformResponse> transform(TransformRequest request);
}

File: clients/go/transport.go
package sharptown

import (
	"context"
	"net/http"
	"time"
)

// Request is the canonical, protocol-independent input handed to a Transport.
type Request struct {
	BaseURL     string
	Headers     map[string]string
	Input       Input
	Operations  map[string]any
	Attachments [][]byte
	Timeout     time.Duration
	HTTPClient  *http.Client
}

// Transport speaks one of the Sharptown protocols (REST, JSON-RPC, gRPC). Each
// implementation accepts the same Request and returns the same Response.
type Transport interface {
	Transform(ctx context.Context, req *Request) (*Response, error)
}

// REST returns the default transport: a multipart POST to {baseURL}/api/v1/transform.
func REST() Transport { return &restTransport{path: "/api/v1/transform", field: "image"} }

// JSONRPC returns the JSON-RPC over WebSocket transport (image.transform at {baseURL}/rpc).
func JSONRPC() Transport { return &jsonrpcTransport{path: "/rpc", method: "image.transform"} }

// GRPC returns the gRPC transport (sharptown.v1.ImageProcessor/Transform, bidi streaming).
func GRPC() Transport { return &grpcTransport{} }
